File: SHELL/fifo.c
#include <stdio.h>
#include <stdlib.h>

#define len(array) sizeof(array) / sizeof(array[0])

enum Bool {
   TRUE  = 1,
   FALSE = 2
};

int refString[] = {
   7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3, 2, 1, 2, 0, 1, 7, 0, 1
};
;

void
printPageStack(int pageStack[], int pageStackSize) {
   for ( int i = 0; i < pageStackSize; i++ ) {
      if ( pageStack[i] != -1 ) {
         fprintf(stdout, "%d ", pageStack[i]);
      }
   }

   fprintf(stdout, "\n");
}

enum Bool
isPresent(int value, int array[], int arraySize) {
   for ( int i = 0; i < arraySize; i++ ) {
      if ( value == array[i] ) {
         return TRUE;
      }
   }

   return FALSE;
}

int
seekIndex(int value, int array[], int arraySize) {
   for ( int i = 0; i < arraySize; i++ ) {
      if ( value == array[i] ) {
         return i;
      }
   }

   return -1;
}

int
fifo(int frames, int refString[], int refStringSize) {
   int  faults       = 0;
   int  replaceCount = 0;
   int* pageStack;

   pageStack = (int*) malloc(frames * sizeof(int));

   for ( int i = 0; i < frames; i++ ) {
      pageStack[i] = -1;
   }

   // fprintf(stdout, "Page Stack Status: \n");

   for ( int i = 0; i < refStringSize; i++ ) {
      if ( isPresent(refString[i], pageStack, frames) == FALSE ) {
         faults++;

         if ( isPresent(-1, pageStack, frames) == TRUE ) {
            pageStack[seekIndex(-1, pageStack, frames)] = refString[i];
         } else {
            pageStack[replaceCount] = refString[i];
            replaceCount            = (replaceCount + 1) % frames;
         }
      }

      // fprintf(stdout, "%d: ", refString[i]);
      // printPageStack(pageStack, frames);
   }

   free(pageStack);

   return faults;
}

int
main(void) {
   fprintf(stdout, "Reference String: ");

   for ( int i = 0; i < len(refString); i++ ) {
      fprintf(stdout, "%d ", refString[i]);
   }

   fprintf(stdout, "\n");

   for ( int i = 1; i < len(refString); i++ ) {
      fprintf(stdout, "Faults caused by %d frames: %d\n", i,
              fifo(i, refString, len(refString)));
   }

   return 0;
}

File: SHELL/fifo.py
ref_string: list = [7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3, 2, 1, 2, 0, 1, 7, 0, 1]


def get_data() -> tuple(int, list):
    frames: int = int(input("Enter number of frames: "))
    ref_string: list(int) = list(map(int, input("Enter reference string: ").split()))
    return frames, ref_string


def print_page_stack(page_stack: list) -> None:
    for _, page in enumerate(page_stack):
        if page != -1:
            print(f"{page} ", end="")

    print()


def fifo(frames: int, ref_string: list) -> int:
    page_stack: list(int) = [-1 for _ in range(frames)]
    faults: int = 0
    replace_count: int = 0

    print("Page Stack Status: ")

    for _, page in enumerate(ref_string):
        try:
            page_stack.index(page)
        except ValueError:
            faults += 1

            if page_stack.count(-1) > 0:
                page_stack[page_stack.index(-1)] = page
            else:
                page_stack[replace_count] = page
                replace_count = (replace_count + 1) % frames

        print(f"{page}: ", end="")
        print_page_stack(page_stack)

    return faults


def main() -> None:
    print(f"Reference String: ", end="")

    for _, page in enumerate(ref_string):
        print(f"{page} ", end="")

    print()

    for frames in range(3, 6):
        print(
            f"Faults caused by {frames} frames: "
            + f"{fifo(frames=frames, ref_string=ref_string)}\n"
        )


if __name__ == "__main__":
    main()
